### PR TITLE
date_time: Add helper functions to boost::date_time::time_duration cl…

### PR DESCRIPTION
**Pull request by gawain-bolton**
…ass.

Currently the time_duration class has a function is_positive() to
determine whether or not the duration is positive.  However, it is
also often useful to know whether or not the duration is negative or
zero.  The is_negative() and is_zero() functions are provided for this
purpose.

Finally, the abs() function returns the absolute value of the duration
which can be useful when checking differences between durations.

### CODE REVIEW

File: include/boost/date_time/time_duration.hpp [hunk 1]
```diff
@@ -29,8 +29,8 @@ namespace date_time {
       This design allows the subclass duration types to provide custom
       construction policies or other custom features not provided here.
 
-      @tparam T The subclass type
-      @tparam rep_type The time resolution traits for this duration type.
+      @param T The subclass type
```
Comment (JeffGarland) [inline include/boost/date_time/time_duration.hpp]: hmm I believe these where just change to tparam in another pull request...

File: include/boost/date_time/time_duration.hpp [hunk 2]
```diff
@@ -144,10 +144,26 @@ namespace date_time {
     {
       return duration_type(ticks_ * (-1));
     }
+    duration_type abs() const
```
Comment (JeffGarland) [inline include/boost/date_time/time_duration.hpp]: These functions seem like a nice addition.  Unfortunately no one will no about them without documentation.  The docs are in boost book xml which is a bit clunky to modify, but it's pretty obvious.  You should be able to find the accessor table for time_duration add trivially add these.

File: test/posix_time/testduration.cpp [hunk 3]
```diff
@@ -17,28 +17,34 @@
...
-  
+
   // construct from components
   time_duration td1(1,25,0);
   time_duration td3(td1.hours(),td1.minutes(),td1.seconds());
   check("total up elements", td1 == td3);
   td1 = -td1; // td1 == "-1:25:00"
+  check("invert_sign",td3 == td1.invert_sign());
+  check("abs",td3 == td1.abs());
+  check("is_positive",td3.is_positive());
```
Comment (JeffGarland) [inline test/posix_time/testduration.cpp]: perhaps add an abs test case for a positive duration so that the coverage checking doesn't complain.

Comment (JeffGarland) [inline test/posix_time/testduration.cpp]: And an is_positive for the false case.  You can see the unit test coverage is flagging missed paths.

**Comment by gawain-bolton** [comment]
Thank you @JeffGarland for your comments.
I have added unit tests and documentation as you suggested.

File: include/boost/date_time/time_duration.hpp (hunk 1 above)
Comment (jeking3) [reply include/boost/date_time/time_duration.hpp]: They were, so this should change to tparam.

File: include/boost/date_time/time_duration.hpp [hunk 4]
```diff
@@ -144,10 +144,26 @@ namespace date_time {
...
+      {
+        return invert_sign();
+      }
+      return duration_type(ticks_);
+    }
     bool is_negative() const
     {
       return ticks_ < 0;
     }
+    bool is_zero() const
```
Comment (jeking3) [inline include/boost/date_time/time_duration.hpp]: operator bool perhaps?  Just a thought.

**Comment by swatanabe** [comment]
AMDG

On 3/5/19 11:08 AM, James E. King III wrote:
> jeking3 requested changes on this pull request.
> 
> Thanks for adding the documentation; the param vs. tparam comment is still a regression, and I was just curious if folks thought an explicit operator bool would be better than is_zero()... ?
> 

I'm against explicit operator bool.

In Christ,
Steven Watanabe

**Comment by JeffGarland** [comment]
> AMDG
> On 3/5/19 11:08 AM, James E. King III wrote: jeking3 requested changes on this pull request. Thanks for adding the documentation; the param vs. tparam comment is still a regression, and I was just curious if folks thought an explicit operator bool would be better than is_zero()... ?
> I'm against explicit operator bool. In Christ, Steven Watanabe

I agree with Steven on this -- this isn't a pointer, iterator, or an optional.  In fact if we were to add operator bool I would expect the behavior to be equivalent to checking for not_a_date_time.

**Comment by gawain-bolton** [comment]
@jeking3: The tparam vs param problem was fixed - I can see no differences between the boost develop branch and my forked develop branch.

@ all:
The travis-ci build failure is due to the fact that the testmicrosec_time_clock.cpp test can fail randomly.
If you execute the test repeatedly, you will only have to wait a few minutes before you get 2 test failures.

After looking at this unit test, it is not clear to me that is has much value.   It also seems to be incorrectly testing the equality of minutes twice instead of testing the equality of minutes and then seconds.

**Comment by jeking3** [comment]
Fair enough, just thought I would ask. :)

**Comment by JeffGarland** [comment]
> @ all:
> The travis-ci build failure is due to the fact that the testmicrosec_time_clock.cpp test can fail randomly.

Hmm what I found was:

terminate called after throwing an instance of 'boost::archive::archive_exception'

Which still would seem unrelated to the change.

> If you execute the test repeatedly, you will only have to wait a few minutes before you get 2 test failures.
> 
> After looking at this unit test, it is not clear to me that is has much value. It also seems to be incorrectly testing the equality of minutes twice instead of testing the equality of minutes and then seconds.

That test has always been an issue, but it is kinda there to help evaluate how monotonic the clock is.  I'd have to look at it in detail again, but clearly that shouldn't hold up this change.

**Comment by jeking3** [comment]
REMINDER (TO MYSELF): SQUASH

**Pull request closed without merging**